Calculate Chassis velocities using pseudo-inverse using normal equations and Cholesky decomposition

JulianBarossi opened the pull request:
chassis_velocities [calc_estimated_chassis_velocities] function

Commit:
Calculate Chasis velocities usingp seudo-inverse using normal equations and cholesky decomposition

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,117 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+
+    for (size_t i = 0; i < n; i++){
+
+      const auto& module = *p_modules[i];
```
we use east const so it would be `auto const& var_name`

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,117 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+
+    float a [3][3] = {{0,0,0},{0,0,0},{0,0,0}};
+    float c [3] = {0,0,0};
+    float x_solution [3] = {0,0,0};
```
you can save you self some work by passing in `{}` the compiler will initialize unspecified values to the default value so you don't need to type out all the zeros.

JulianBarossi replied:
Ya I figured, but it was more for visualization purposes.

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,117 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+    for (size_t i = 0; i < n; i++){
+      accumulate(1.0f, 0.0f, -y[i], vix[i]);
+      accumulate(0.0f, 1.0f, x[i], viy[i]);
+    }
```
it seams to me that you can't accumulate as you get each value. `x`,`y`,`vix`, and `viy`'s elements are all just values gotten from the module and never changed. If you calculate the element value at accumulation time then it would get rid of 4 extra arrays.

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,117 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+
+    if (!cholesky(a,c,x_solution)){
+      return results;
+    }
```
What is Cholesky checking? (if it is checking validity of a calculation then throw an error don't return a velocity of 0)

MichaelYKersey reviewed (commented):
I will trust that the actual math correct, but add some comments so someone has some ideas of what to search up/look at if they want to understand in the future.

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,117 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+
+    auto accumulate = [&](float r0, float r1, float position, float v){
+
+      //1st coloum
```
column is spelled wrong in a couple places

JulianBarossi replied:
ya I dont like english

Commit:
calculate interpolation time

Commit:
added some references for normal equations least squares with cholesky decomposition

changed commit to the correct branch

Commit:
Updates

-Removed temp arrays and accumulate directly now

-Converted angles to rad and stored angular velocity as deg/s

-Cholesky check now returns false when it fails computational check

MichaelYKersey commented on drive/include/drivetrain_math.hpp:
```diff
@@ -141,4 +141,6 @@ std::array<swerve_module_state, module_count> interpolate_states(
  */
 float modulus_range(float p_value, float p_lower, float p_upper);
+
+bool isSafe(const std::array<swerve_module_state, module_count>& modules)
```
use `type const&` not `const type&`

Commit:
fixed return type for chassis velocities and added isSafe check for interpolation

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,118 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+// https://www.youtube.com/watch?v=C7LEuhS4H94&t=8s
+// https://math.mit.edu/icg/resources/teaching/18.085-spring2015/LeastSquares.pdf
+// https://tobydriscoll.net/fnc-julia/leastsq/normaleqns.html
+constexpr float M_PI = 3.14159265358979323846f;
```
c++ lib has it's own pi const, use that

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -48,9 +48,118 @@ std::array<vector2d, module_count> chassis_velocities_to_module_vectors(
+    float const xi = module.settings.position.x;
+    float const yi = module.settings.position.y;
+
+    auto const state = module.get_actual_state_cache();
+    float const speed = state.propulsion_velocity;
+    float const angle_degrees = state.steer_angle;
+    float const angle_rad = angle_degrees * deg_to_rad;
+    float const vix = speed * std::cos(angle_rad);
+    float const viy = speed * std::sin(angle_rad);
```
might be good to make a state to velocity vector function if it doesn't already exist

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -206,4 +315,82 @@ float modulus_range(float p_value, float p_lower, float p_upper)
+  constexpr float deg_to_rad = M_PI / 180.0f;
+
+  for (size_t i = 0; i < module_count; i++) {
+    auto const& module = *p_modules[i];
+    auto const state = module.get_actual_state_cache();
+    auto const& position = module.settings.position;
+
+    // convert deg/sec to rad/sec
+    float omega_radians = estimates.rotational_vel * deg_to_rad;
+
+    float est_vx = estimates.translation.x - omega_radians * position.y;
+    float est_vy = estimates.translation.y + omega_radians * position.x;
```
these math calculations (module velocity vector & chassis speeds to module sate or vector) should already exist in some other drive math function, if not then make one.

MichaelYKersey commented on drive/src/drivetrain_math.cpp:
```diff
@@ -206,4 +315,82 @@ float modulus_range(float p_value, float p_lower, float p_upper)
+
+// helper to find difference given two angels
+// returns radians
+float angleDiff(float a, float b)
+{
+
+  float d = std::fmod(a - b + M_PI, 2.0 * M_PI);
+  if (d < 0)
+    d += 2.0 * M_PI;
+
+  return d - M_PI;
+}
```
Is there a reason to noy use `modulus_range`

MichaelYKersey reviewed (commented):
make a simple print demo that runs a couple test cases while I get unit tests set up to make sure the code works. Also rebase from main.
